tests/kafkatest/services/kafka/kafka.py

```diff
@@ -56,24 +72,34 @@ def start(self):
                 self.create_topic(topic_cfg)
 
     def start_node(self, node):
-        props_file = self.render('kafka.properties', node=node, broker_id=self.idx(node))
+        cfg = KafkaConfig(**node.config)
+        cfg[kafka_prop.ADVERTISED_HOSTNAME] = node.account.hostname
+        cfg[kafka_prop.ZOOKEEPER_CONNECT] = self.zk.connect_setting()
+        props_file = cfg.render()
         self.logger.info("kafka.properties:")
         self.logger.info(props_file)
         node.account.create_file("/mnt/kafka.properties", props_file)
 
-        cmd = "/opt/kafka/bin/kafka-server-start.sh /mnt/kafka.properties 1>> /mnt/kafka.log 2>> /mnt/kafka.log & echo $! > /mnt/kafka.pid"
+        cmd = "export LOG_DIR=/mnt/kafka-operational-logs/; "
+        cmd += self._kafka_dir(node) + "/bin/kafka-server-start.sh /mnt/kafka.properties 1>> /mnt/kafka.log 2>> /mnt/kafka.log &"
         self.logger.debug("Attempting to start KafkaService on %s with command: %s" % (str(node.account), cmd))
         with node.account.monitor_log("/mnt/kafka.log") as monitor:
             node.account.ssh(cmd)
             monitor.wait_until("Kafka Server.*started", timeout_sec=30, err_msg="Kafka server didn't finish startup")
-        if len(self.pids(node)) == 0:
-            raise Exception("No process ids recorded on node %s" % str(node))
+
+    def _kafka_dir(self, node):
+        if node.version == KafkaVersion.TRUNK:
+            return "/opt/kafka"
+        else:
+            return "/opt/kafka-" + node.version
 
     def pids(self, node):
         """Return process ids associated with running processes on the given node."""
         try:
-            return [pid for pid in node.account.ssh_capture("cat /mnt/kafka.pid", callback=int)]
-        except:
+            cmd = "ps ax | grep -i kafka.properties | grep java | grep -v grep | awk '{print $1}'"
```

Comment:
I don't think pid files are brittle at all. They have a few well known failure modes (one obvious one being if a pid is reused then you can kill the wrong process). But they are very widely used for a reason. They are a proven, reliable approach to process management. This ps + grep process is *not* a normal way of managing processes via scripts -- people often use it as a handy way of doing ad hoc process management, but that's pretty much it. I can't think of any other project that uses this approach.

This `ps` + `grep` + `kill` approach is probably least problematic in the context of ducktape because we force each service to have a dedicated node. It fails utterly if you ever want to be able to run multiple instances of a service on a single node...

Most of the uses are holdovers from the fact that Kafka's bin/ scripts were written this way. It was convenient to get services up and running by just reusing that code, but that doesn't mean that's what we *should* do. I was just surprised to see that the patch original code had pid files and for some reason we were switching back.

By the way, there are a bunch of other things that a good service init script would handle that Kafka's don't, and we're working around them. Normal service scripts would validate that the process started properly, returning a useful exit code if the service failed quickly.